### PR TITLE
[[FIX]] Support property assignment when destructure assigning

### LINKED ISSUE #2659: array destructuring assignment to an object property errors incorrectly
Regression in 2.9

```
let a = {};

[a.b] = ['b'];
```

yields

```
Expected ',' and saw '.'
```

### LINKED ISSUE #2660: array destructuring assignment to an object property where the object is a const errors incorrectly
Regression is 2.9

```
const a = {};

[a.b] = ['b'];
```

yields:
```
Attempting to override 'a' which is a constant.
```

### PR DESCRIPTION
**Pull request by lukeapage**
Fixes #2659 and fixes #2660

### CODE REVIEW

**Comment by lukeapage** [comment]
comments welcome (and ideas for any more tests) - I got it working but I don't know if there is a better way.

**Comment by nicolo-ribaudo** [comment]
This code is allowed by this pr, while it shouldn't:
```js
let a;
[a()] = [];
```

**Comment by lukeapage** [comment]
Well spotted.

> What do you think about using checkLeftSideAssign?

I've fixed the bug by setting the precedence to 155 rather than 150 and I've verified no other operators have a precedence higher. But I think you are right, there are checks in `checkLeftSideAssign` that should be being run. I'll handle this next.

**Comment by lukeapage** [comment]
> But I think you are right, there are checks in `checkLeftSideAssign` that should be being run. I'll handle this next.

done - there were a couple of bugs because of this even in master (e.g. `({e} = {});` in a try-catch).

**Comment by mikesherov** [comment]
:+1: AFAICT

File: tests/unit/parser.js [hunk 1]
```diff
@@ -1643,7 +1646,48 @@ exports["destructuring const errors"] = function (test) {
...
+    .addError(5, "Missing semicolon.")
+    .addError(5, "Expected an identifier and instead saw ']'.")
+    .addError(5, "Expected an assignment or function call and instead saw an expression.")
+    .addError(5, "Missing semicolon.")
+    .test(code, {esnext: true, unused: true, undef: true});
+
+  test.done();
+};
+
+exports["destructuring const errors"] = function (test) {
```
Comment (jugglinmike) [inline tests/unit/parser.js]: Looks like some weird copy-paste behavior might have tripped you up here--this test is identical to the one that precedes it. It's safe to remove.

File: src/jshint.js [hunk 2]
```diff
@@ -3374,9 +3380,27 @@ var JSHINT = (function() {
...
+          var identifierToken = is_rest ? peek(0) : state.tokens.next;
+          if (!identifierToken.identifier) {
+            warning("E030", identifierToken, identifierToken.value);
+          }
+          var assignTarget = expression(155);
+          if (assignTarget) {
+            checkLeftSideAssign(assignTarget);
+
+            // if the target was a simple identifier, add it to the list to return
+            if (assignTarget.identifier) {
```
Comment (jugglinmike) [inline src/jshint.js]: The converse of this condition is not currently tested. In other words, this also passes:

```diff
-// if the target was a simple identifier, add it to the list to return
-if (assignTarget.identifier) {
   ident = assignTarget.value;
-}
```

...but that's clearly wrong. I'm having trouble thinking of a test case for that, though. Any ideas?

**Comment by lukeapage** [comment]
> The converse of this condition is not currently tested. In other words, this also passes:

oh,  really? for `[ a.b] = [] ` isnt assignTarget the '.' with a left and right identifier? maybe value is not set on it which means it would be ignored,  explaining why it passes tests with if commented out? 

I've been trying not to return null identifiers from this function but still guarding against them. 

I will debug and double check tomorrow, just mentioning in case this explains something for you.

File: tests/unit/parser.js (hunk 1 above)
Comment (lukeapage) [inline tests/unit/parser.js]: woops,  I was going to split the test but I changed my  mind.  looks like i didn't undo far enough

**Comment by caitp** [comment]
I haven't read this yet, but a good test case ( for both destructuring assignment and vanilla assignment) is `new.target` for assignment target. Definitely want to warn about that even though it is an early error in VMs

**Comment by lukeapage** [comment]
yep testing that :)

File: tests/unit/parser.js (hunk 1 above)
Comment (lukeapage) [inline tests/unit/parser.js]: fixed.

**Comment by lukeapage** [comment]
So, if I change the code to

```
// if the target was a simple identifier, add it to the list to return
if (assignTarget.identifier) {
    ident = assignTarget.value;
} else {
    console.log("not an identifier");
    console.log("value =", assignTarget.value);
}
```

and run the tests I get lots of 

```
not an identifier
value = .
not an identifier
value = .
not an identifier
value = [
not an identifier
value = [
not an identifier
value = ...
```

which initially confused me because if I remove the if, things still pass.. I would expect lots of `'[' is not defined.`. 
However the usage of that identifier list only does
```
if (t.id) {
    state.funct["(scope)"].block.modify(t.id, t.token);
}
```
and modify doesn't "use" the variable - that is done when an identifier is parsed.

So I've added a couple of tests that make sure we get a `'x' is not defined` warning when destruct assigning and that `[...i] = [1]` uses `i`.

**Comment by jugglinmike** [comment]
Let me see if I'm understanding this correctly: due to the way the `destructuringPattern` function's return value is used, it currently makes no difference if it erroneously includes entries for non-identifier tokens. Is that right?

**Comment by lukeapage** [comment]
yes.

I could change that by making scope managers modify throw an error if it is
not first used, but I'm reluctant to add an assertion unless it only
asserts during tests (in case it introduces a bug).

**Comment by jugglinmike** [comment]
Yeah, I'd be leery of bugs and also of the assertion's effect on performance. We should look into a build step for the main release--that would let us strip out assertions (and it would also pave the way for other transformations, like ES6 syntax).

That's work for another day, though. I've squashed these commits into one and merged that to `master` at revision b6df1f22633ee597df433591b0fcf925c98801f0. Thanks, Luke!

**Pull request closed without merging**